### PR TITLE
feat: Add recordEvent API and expose Plugin to enable recording of custom events

### PR DESCRIPTION
**Pull request by adebayor123**
## Detail
This PR exposes the Plugin Interface and two new APIs, `recordEvent` and `addPlugin`, to allow customers to implement customized plugins and record custom event using the web client. With this change, customers can record custom events by either making a direct API call to record events or implementing and registering their customized plugins.

Closes #185 and #139 

## Testing
Added new unit test cases and integ test cases to verify behavior.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/custom_events.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Using CloudWatch RUM Web Client to Record Custom Events
```
Comment (adebayor123) [inline docs/custom_events.md:1]: Thought a dedicated document would be better than appending onto `docs/cdn_commands.md`. Let me know if you think otherwise.

File: src/orchestration/Orchestration.ts [hunk 2]
```diff
@@ -352,6 +353,17 @@ export class Orchestration {
         );
     }
 
+    /**
+     * Record an event directly to eventCache.
+     * @param type
+     * @param eventData
+     */
+    public recordEvent(eventType: string, eventData: object) {
+        if (!EVENT_TYPE_NAMESPACE.has(eventType)) {
```
Comment (adebayor123) [inline src/orchestration/Orchestration.ts:362]: Removing this check allows users to send built-in events by creating their own data (assuming it conforms to the schema), which addresses a use case brought up before, and in the case where they do send wrong data format, at the end of the day, it will be handled by the Data Plane. Let me know what you think.

File: src/event-cache/EventCache.ts [hunk 3]
```diff
@@ -88,7 +91,11 @@ export class EventCache {
             const session: Session = this.sessionManager.getSession();
             this.sessionManager.incrementSessionEventCount();
 
-            if (this.canRecord(session)) {
+            if (
+                this.canRecord(session) &&
+                this.isEventTypePatternAllowed(type) &&
+                Object.keys(eventData).length !== 0
```
Comment (adebayor123) [inline src/event-cache/EventCache.ts:97]: Since `eventCache.recordEvent` was only accessible via internal implementations, we would never have to worry about empty data. Now that customers can directly send in data and record "custom" events, it is possible that empty data would be passed in, and I thought the benefit of preventing the ingestion of empty data outweighed the use case of sending empty data.

File: src/CommandQueue.ts [hunk 4]
```diff
@@ -47,6 +56,20 @@ export class CommandQueue {
...
             this.orchestration.registerDomEvents(payload);
         },
+        recordEvent: (payload: RecordEventInput) => {
+            const type = payload.event_type;
+            const eventData = payload.event_data;
+            if (
+                type === undefined ||
+                eventData === undefined ||
+                typeof type !== 'string' ||
+                typeof eventData !== 'object'
```
Comment (maniator) [inline src/CommandQueue.ts:66]: Can we make event data _not_ required?

Sometimes we may want to record a straight event with no data

File: src/event-cache/EventCache.ts (hunk 3 above)
Comment (maniator) [reply src/event-cache/EventCache.ts:97]: Why are we preventing empty data?

Sometimes we might just want to know how many times one event happened (with no data attached to that event)

File: src/orchestration/__tests__/Orchestration.test.ts [hunk 5]
```diff
@@ -6,6 +6,7 @@ import { DomEventPlugin } from '../../plugins/event-plugins/DomEventPlugin';
 import { JsErrorPlugin } from '../../plugins/event-plugins/JsErrorPlugin';
 import { PluginManager } from '../../plugins/PluginManager';
 import { PageIdFormatEnum } from '../Orchestration';
+import { JS_ERROR_EVENT_TYPE } from '../../plugins/utils/constant';
```
Comment (maniator) [inline src/orchestration/__tests__/Orchestration.test.ts:9]: why is this added?

File: src/plugins/utils/constant.ts [hunk 6]
```diff
@@ -28,3 +26,20 @@ export const PAGE_VIEW_EVENT_TYPE = `${RUM_AMZ_PREFIX}.page_view_event`;
 
 // Session start event
 export const SESSION_START_EVENT_TYPE = `${RUM_AMZ_PREFIX}.session_start_event`;
+
+// Protected namespace for built-in events
+export const EVENT_TYPE_NAMESPACE = new Set<string>([
```
Comment (maniator) [inline src/plugins/utils/constant.ts:31]: You may want to wrap this in `Object.freeze` otherwise it is mutable.

File: src/CommandQueue.ts [hunk 7]
```diff
@@ -21,6 +22,14 @@ export type AwsRumClientInit = {
     u?: string;
 };
 
+/**
+ * The input for recordEvent command
+ */
+export type RecordEventInput = {
+    event_type: string;
+    event_data: object;
```
Comment (maniator) [inline src/CommandQueue.ts:30]: can we make the data optional?

Comment (maniator) [reply src/CommandQueue.ts:30]: (see other comments)

File: app/custom_event.html [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            function recordEvent() {
+                cwr('recordEvent', {
+                    event_type: 'custom_event_api',
+                    event_data: { customEventVersion: 255 }
+                });
+            }
+
+            // Add plugin
+            function addPlugin() {
+                class newPlugin {
```
Comment (maniator) [inline app/custom_event.html:28]: Can we use js paradigm of naming classes with uppercase letters?

File: app/custom_event.html [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                    }
+                    getPluginId() {
+                        return this.id;
+                    }
+                    record(data) {
+                        this.context.record(this.id, data);
+                    }
+                }
+                this.plugin = new newPlugin();
+                cwr('addPlugin', this.plugin);
```
Comment (maniator) [inline app/custom_event.html:51]: shouldn't `addPlugin` make sure that the plugin is an instance of the plugin `Plugin` class?

Comment (maniator) [reply app/custom_event.html:51]: oooo we have it as an interface...

We may want to change that to make sure passed in plugins are extending the plugin class type (https://github.com/aws-observability/aws-rum-web/blob/main/src/plugins/Plugin.ts)

File: docs/cdn_commands.md [hunk 10]
```diff
@@ -61,7 +61,8 @@ if(awsCreds) awsRum.setAwsCredentials(credentialProvider);
...
 | setAwsCredentials | [Credentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Credentials.html) \|&nbsp;[CredentialProvider](https://www.npmjs.com/package/@aws-sdk/credential-providers) | `cwr('setAwsCredentials', cred);`<br/><br/>`awsRum.setAwsCredentials(cred);` | Forward AWS credentials to the web client. The web client requires AWS credentials with permission to call the `PutRumEvents` API. If you have not set `identityPoolId` and `guestRoleArn` in the web client configuration, you must forward AWS credentials to the web client using this command.
-
+| recordEvent | Object | `cwr('recordEvent', {event_type: 'your_event_type', event_data: {field1: 1, field2: 2}})` <br/><br/> `awsRum.recordEvent('your_event_type', {field1 1, field2: 2})` | Record an event. Learn more in [Using CloudWatch RUM Web Client to Record Custom Events](custom_events.md)
+| addPlugin | Plugin | `cwr('addPlugin', this.myCustomPlugin)` <br/><br/> `awsRum.addPlugin(this.myCustomPlugin)` | Adds a plugin to the Web Client. Learn more in [Using CloudWatch RUM Web Client to Record Custom Events](custom_events.md)
```
Comment (qhanam) [inline docs/cdn_commands.md:65]: I'm probably contradicting something I said earlier, but it might be better to do docs in a separate commit. This way we can separate the technical conversation from the doc revisions.

File: src/CommandQueue.ts [hunk 11]
```diff
@@ -47,6 +56,20 @@ export class CommandQueue {
         registerDomEvents: (payload: any): void => {
             this.orchestration.registerDomEvents(payload);
         },
+        recordEvent: (payload: RecordEventInput) => {
+            const type = payload.event_type;
+            const eventData = payload.event_data;
+            if (
+                type === undefined ||
+                eventData === undefined ||
```
Comment (qhanam) [inline src/CommandQueue.ts:64]: nit: These checks are redundant.
```
typeof undefined === 'undefined'
```

File: src/CommandQueue.ts [hunk 12]
```diff
@@ -47,6 +56,20 @@ export class CommandQueue {
         registerDomEvents: (payload: any): void => {
             this.orchestration.registerDomEvents(payload);
         },
+        recordEvent: (payload: RecordEventInput) => {
+            const type = payload.event_type;
+            const eventData = payload.event_data;
```
Comment (qhanam) [inline src/CommandQueue.ts:61]: nit: Using `payload.type` and `payload.data` is fine, no need to re-assign.

File: src/CommandQueue.ts [hunk 13]
```diff
@@ -47,6 +56,20 @@ export class CommandQueue {
         registerDomEvents: (payload: any): void => {
             this.orchestration.registerDomEvents(payload);
         },
+        recordEvent: (payload: RecordEventInput) => {
```
Comment (qhanam) [inline src/CommandQueue.ts:59]: issue: If we're going to check the payload for type safety, the proper type is `unknown` (indicates the type could be anything but is the opposite of `any` in that `unknown` values cannot be assigned to any other type).

File: src/CommandQueue.ts [hunk 14]
```diff
@@ -47,6 +56,20 @@ export class CommandQueue {
...
+            const type = payload.event_type;
+            const eventData = payload.event_data;
+            if (
+                type === undefined ||
+                eventData === undefined ||
+                typeof type !== 'string' ||
+                typeof eventData !== 'object'
+            ) {
+                throw new Error('IncorrectParametersException: invalid input.');
```
Comment (qhanam) [inline src/CommandQueue.ts:68]: nit: For consistency.
```suggestion
                throw new Error('IncorrectParametersException')
```

File: src/event-cache/__tests__/EventCache.integ.test.ts [hunk 15]
```diff
@@ -25,9 +25,9 @@ describe('EventCache tests', () => {
         const eventCache: EventCache = Utils.createEventCache(config);
 
         // Run
-        eventCache.recordEvent(EVENT1_SCHEMA, {});
+        eventCache.recordEvent(EVENT1_SCHEMA, { field: 'value' });
```
Comment (qhanam) [inline src/event-cache/__tests__/EventCache.integ.test.ts:28]: nit: Once we revert to allowing empty objects in `recordEvent`, please revert these :-)

File: src/orchestration/Orchestration.ts (hunk 2 above)
Comment (adebayor123) [reply src/orchestration/Orchestration.ts:362]: Sounds good, as it also allows users to send built-in events on demand. Will reflect in next revision.

File: src/event-cache/__tests__/EventCache.integ.test.ts (hunk 15 above)
Comment (adebayor123) [reply src/event-cache/__tests__/EventCache.integ.test.ts:28]: Done!

File: src/orchestration/Orchestration.ts [hunk 16]
```diff
@@ -352,6 +353,17 @@ export class Orchestration {
         );
     }
 
+    /**
+     * Record an event directly to eventCache.
+     * @param type
+     * @param eventData
+     */
+    public recordEvent(eventType: string, eventData: object) {
```
[the remark this answers is not in the document]
Comment (adebayor123) [reply src/orchestration/Orchestration.ts:361]: I think object is better due to two reasons:
1. We will serialize in `eventCache.addRecordToCache`
2. Accepting object will lower the chances of users sending invalid JSON objects in the `details` field.

File: app/custom_event.html (hunk 8 above)
Comment (adebayor123) [reply app/custom_event.html:28]: Agreed!

File: src/CommandQueue.ts (hunk 11 above)
Comment (adebayor123) [reply src/CommandQueue.ts:64]: Good suggestion

File: src/event-cache/EventCache.ts (hunk 3 above)
Comment (adebayor123) [reply src/event-cache/EventCache.ts:97]: I admit, while implementing this I was mistaking empty data with null values. I'll be making `{}` as default value if not passed in.

File: src/CommandQueue.ts (hunk 4 above)
Comment (adebayor123) [reply src/CommandQueue.ts:66]: Updated CommandQueue to be able to send events without containing eventData field, and Orchestration to give default value of empty object when not passed in.

File: src/event-cache/EventCache.ts [hunk 17]
```diff
@@ -228,4 +235,12 @@ export class EventCache {
 
         return include && !exclude;
     }
+
+    /**
+     * Returns true if the eventType consists of valid characters. Returns false otherwise.
+     */
+    private isEventTypePatternAllowed(eventType: string) {
+        const valid = new RegExp('^(?!\\.)[\\.\\-_#A-Za-z0-9]+$');
```
[the remark this answers is not in the document]
Comment (adebayor123) [reply src/event-cache/EventCache.ts:243]: (1) These are the characters that are allowed for AppMonitor names, so it somewhat maintains consistency in naming convention,  (2) the only difference between existing event types and this pattern is the addition of `#, -, _`.

From what I found, it seems there's no so called `invalid` characters for field values in ES, but I think it's better to follow the syntax we have for AppMonitor names.

File: app/custom_event.html (hunk 9 above)
Comment (adebayor123) [reply app/custom_event.html:51]: It seems the best way to check if an object implements an interface is to manually do user defined type guard (checking if the object has access to all the methods), which means any Object that has the methods in the Interface will pass the type guard. Do you have suggestions for this?

File: docs/custom_events.md (hunk 1 above)
Comment (adebayor123) [inline docs/custom_events.md:1]: Add example document that shows how to do x,y,z

File: src/event-cache/EventCache.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (adebayor123) [reply src/event-cache/EventCache.ts:243]: Synced up offline, decided to go without the pattern check

File: app/custom_event.html (hunk 9 above)
Comment (adebayor123) [reply app/custom_event.html:51]: Synced up with Quinn, whether we add the type check or not, the plugin will most likely fail during `addPlugin` stage.

File: src/CommandQueue.ts [hunk 18]
```diff
@@ -65,6 +80,9 @@ export class CommandQueue {
             } else {
                 throw new Error('IncorrectParametersException');
             }
+        },
+        addPlugin: (plugin: Plugin) => {
+            this.orchestration.addPlugin(plugin);
```
Comment (qhanam) [inline src/CommandQueue.ts:85]: issue: We have an interface for adding plugins in the [`AwsRumConfiguration` object](https://github.com/aws-observability/aws-rum-web/blob/main/src/orchestration/Orchestration.ts#L72) ([example usage](https://github.com/aws-observability/aws-rum-web/blob/main/src/loader/loader-js-error-event.js#L8)). 

Is there benefit for having this as well?

File: src/event-cache/EventCache.ts [hunk 19]
```diff
@@ -75,11 +75,11 @@ export class EventCache {
      * Add an event to the cache and reset the session timer.
      *
      * If the session is being recorded, the event will be recorded.
-     * If the session is not being recorded, the event will not be recorded.
+     * If the session is not being recorded, the event will not be recorded
      *
      * @param type The event schema.
      */
-    public recordEvent = (type: string, eventData: object) => {
+    public recordEvent = (type: string, eventData = {}) => {
```
Comment (qhanam) [inline src/event-cache/EventCache.ts:82]: nit: Missing type

File: src/orchestration/Orchestration.ts [hunk 20]
```diff
@@ -352,6 +352,15 @@ export class Orchestration {
         );
     }
 
+    /**
+     * Records a custom event.
+     * @param type A unique name for the type of event being recorded.
+     * @param eventData A JSON object containing the event's attributes.
+     */
+    public recordEvent(eventType: string, eventData = {}) {
```
Comment (qhanam) [inline src/orchestration/Orchestration.ts:360]: nit: Missing explicit type declaration

File: docs/cdn_commands.md [hunk 21]
```diff
@@ -61,7 +61,8 @@ if(awsCreds) awsRum.setAwsCredentials(credentialProvider);
...
 | setAwsCredentials | [Credentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Credentials.html) \|&nbsp;[CredentialProvider](https://www.npmjs.com/package/@aws-sdk/credential-providers) | `cwr('setAwsCredentials', cred);`<br/><br/>`awsRum.setAwsCredentials(cred);` | Forward AWS credentials to the web client. The web client requires AWS credentials with permission to call the `PutRumEvents` API. If you have not set `identityPoolId` and `guestRoleArn` in the web client configuration, you must forward AWS credentials to the web client using this command.
-
+| recordEvent | Object | `cwr('recordEvent', {event_type: 'your_event_type', event_data: {field1: 1, field2: 2}})` <br/><br/> `awsRum.recordEvent('your_event_type', {field1: 1, field2: 2})` | Record a custom event. Learn more in [Using CloudWatch RUM Web Client to Record Custom Events](custom_events.md)
```
Comment (qhanam) [inline docs/cdn_commands.md:64]: nit: Broken link.

File: src/event-cache/EventCache.ts (hunk 19 above)
Comment (adebayor123) [reply src/event-cache/EventCache.ts:82]: From what I saw, default values declares the type as well: https://www.typescriptlang.org/docs/handbook/functions.html - The reason for adding a default value was due to this comment: https://github.com/aws-observability/aws-rum-web/pull/188#discussion_r937359770

File: src/orchestration/Orchestration.ts (hunk 20 above)
Comment (adebayor123) [reply src/orchestration/Orchestration.ts:360]: Answered above

File: docs/cdn_commands.md (hunk 21 above)
Comment (adebayor123) [reply docs/cdn_commands.md:64]: The link should be broken because the documentations were separated into a different file - #195

File: src/CommandQueue.ts (hunk 18 above)
Comment (adebayor123) [reply src/CommandQueue.ts:85]: Good point. One case that would be beneficial of having an `addPlugin` method is that it gives the users the flexibility to initialize a plugin at any point in time. Initializing plugins only via the config does not allow that as far as I know.

File: src/event-cache/EventCache.ts (hunk 19 above)
Comment (adebayor123) [reply src/event-cache/EventCache.ts:82]: Actually, just tested, it does not declare the type explicitly. Will be making changes

File: src/CommandQueue.ts (hunk 18 above)
Comment (qhanam) [reply src/CommandQueue.ts:85]: Plugins lifecycle management happens through enable, disable and update, which I think covers any situation where `addPlugin` would be used. Since there is no `deletePlugin`, the expectation is that plugins live for the duration of the web client instance.

Comment (maniator) [inline src/CommandQueue.ts:85]: Do we want type checks here to make sure the passed value is a plugin? (not every consumer will be using typescript)

File: app/custom_event.html [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+                cwr('recordEvent', {
+                    event_type: 'custom_event_api',
+                    event_data: { customEventVersion: 255 }
+                });
+            }
+
+            function recordEmptyEvent() {
+                cwr('recordEvent', {
+                    event_type: 'custom_event_api',
+                    event_data: {}
```
Comment (maniator) [inline app/custom_event.html:29]: will you always have to send an empty object if you dont want to send any data?

Comment (adebayor123) [reply app/custom_event.html:29]: Yes, that's the idea.

**Comment by maniator** [comment]
Any updates on when this might get deployed to the next version.

File: src/orchestration/Orchestration.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (maniator) [reply src/orchestration/Orchestration.ts:360]: I would say make it generic:

```typescript
public recordEvent<T extends unknown = object>(eventType: string, eventData: T) {
...

File: src/CommandQueue.ts [hunk 23]
```diff
@@ -61,6 +63,20 @@ export class CommandQueue {
...
         },
+        recordEvent: (payload: any) => {
+            if (
+                typeof payload === 'object' &&
+                typeof payload.event_type === 'string' &&
+                typeof payload.event_data === 'object'
+            ) {
+                this.orchestration.recordEvent(
+                    payload.event_type,
+                    payload.event_data
```
Comment (qhanam) [inline src/CommandQueue.ts:74]: issue: Convention is to use camel case for object property names.

**Review by limhjgrace — APPROVED** [review]
Synced offline to clarify some questions, lgtm!

**Review by qhanam — APPROVED** [review]

File: docs/cdn_commands.md [hunk 24]
```diff
@@ -62,6 +62,7 @@ if(awsCreds) awsRum.setAwsCredentials(credentialProvider);
...
 | setAwsCredentials | [Credentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Credentials.html) \|&nbsp;[CredentialProvider](https://www.npmjs.com/package/@aws-sdk/credential-providers) | `cwr('setAwsCredentials', cred);`<br/><br/>`awsRum.setAwsCredentials(cred);` | Forward AWS credentials to the web client. The web client requires AWS credentials with permission to call the `PutRumEvents` API. If you have not set `identityPoolId` and `guestRoleArn` in the web client configuration, you must forward AWS credentials to the web client using this command.
+| recordEvent | Object | `cwr('recordEvent', {type: 'your_event_type', data: {field1: 1, field2: 2}})` <br/><br/> `awsRum.recordEvent('your_event_type', {field1: 1, field2: 2})` | Record a custom event. Learn more in [Using CloudWatch RUM Web Client to Record Custom Events](examples.md)
```
Comment (qhanam) [inline docs/cdn_commands.md:65]:
```suggestion
| recordEvent | Object | `cwr('recordEvent', {type: 'your_event_type', data: {field1: 1, field2: 2}})` <br/><br/> `awsRum.recordEvent('your_event_type', {field1: 1, field2: 2})` | Record a custom event. 
```
